File: gren/v1/graphs/reducers/write-file.js
// Side-effect reducer: write text (or JSON) to a file. Deterministic file name per run.
// args: { dir: "out", name: "report", ext: "md", field?: "markdown" }   input: { markdown | text | content | ... }
import fs from "node:fs";
import path from "node:path";

export default async function writeFile(input, args, ctx) {
  const dir = path.resolve(String(args.dir ?? "out"));
  const ext = String(args.ext ?? "md");
  const field = args.field ? String(args.field) : undefined;
  const raw = field ? input[field] : (input.markdown ?? input.text ?? input.content ?? input);
  const body = typeof raw === "string" ? raw : JSON.stringify(raw, null, 2);
  const base = String(args.name ?? "report").replace(/[^a-zA-Z0-9_-]+/g, "-").slice(0, 60);
  const stamp = ctx.runId.replace(/[^a-zA-Z0-9_-]+/g, "_").slice(-28);
  const file = path.join(dir, `${base}-${stamp}.${ext}`);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(file, body, "utf8");
  ctx.log(`wrote ${file} (${body.length} chars)`);
  return { written: true, file: file.replace(/\\/g, "/"), bytes: Buffer.byteLength(body), at: new Date().toISOString() };
}

File: gren/kit/graphs/reducers/write-file.py
# Side-effect reducer: write text (or JSON) to a file. Deterministic file name per run.
# args: { dir: "out", name: "report", ext: "md", field?: "markdown" }   input: { markdown | text | content | ... }
import json
import os
import re
import time


def reduce(input, args, ctx):
    d = os.path.abspath(str(args.get("dir", "out")))
    ext = str(args.get("ext", "md"))
    field = str(args["field"]) if args.get("field") else None
    raw = input.get(field) if field else (input.get("markdown") if input.get("markdown") is not None else input.get("text") if input.get("text") is not None else input.get("content") if input.get("content") is not None else input)
    body = raw if isinstance(raw, str) else json.dumps(raw, indent=2, ensure_ascii=False, default=str)
    base = re.sub(r"[^a-zA-Z0-9_-]+", "-", str(args.get("name", "report")))[:60]
    stamp = re.sub(r"[^a-zA-Z0-9_-]+", "_", ctx.run_id)[-28:]
    file = os.path.join(d, f"{base}-{stamp}.{ext}")
    os.makedirs(d, exist_ok=True)
    with open(file, "w", encoding="utf-8") as f:
        f.write(body)
    ctx.log(f"wrote {file} ({len(body)} chars)")
    return {"written": True, "file": file.replace("\\", "/"), "bytes": len(body.encode("utf-8")), "at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())}
